io.openems.edge.controller.symmetric.peakshaving/test/io/openems/edge/controller/symmetric/peakshaving/ControllerEssPeakShavingImplTest.java

```diff
@@ -94,4 +104,111 @@ public void test() throws Exception {
+						.output(ESS_SET_ACTIVE_POWER_EQUALS, 10_000)) //
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 52)//
+						.output(ESS_SET_ACTIVE_POWER_EQUALS, 10_000)) //
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 53)//
+						.output(ESS_SET_ACTIVE_POWER_EQUALS, 10_000)) //
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 54)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000)) //
+				.next(new TestCase("Above peak shaving power") //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 54)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000)) //
+				.next(new TestCase("within boundaries") //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 40_000) //
+						.input(ESS_SOC, 54)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 0)) //
+				.next(new TestCase("below recharge power") //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 20_000) //
+						.input(ESS_SOC, 54)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, -10_000))
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 54)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 53)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 52)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 52)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 51)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 51)//
+						.output(ESS_SET_ACTIVE_POWER_GREATER_THAN_EQUALS, 10_000))
+
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 50)//
+						.output(ESS_SET_ACTIVE_POWER_EQUALS, 10_000))
+
+				.next(new TestCase() //
+						.input(ESS_ACTIVE_POWER, 0) //
+						.input(METER_ACTIVE_POWER, 60_000) //
+						.input(ESS_SOC, 49)//
+						.output(ESS_SET_ACTIVE_POWER_EQUALS, 10_000))
```

Comment:
Is there any good and easy way to test this controller together with another controller?
Our groups best guess was the snippet below, which had one flaw where we did not find an easy fix. The flaw was that the DummyEss battery does not apply any constraints. So we can test that the controller outputs the right values to the channels, but not that follow up controllers actually perform multi use. We've tested it manually by setting up a simulation environment instead.

```
		var ess = new DummyManagedSymmetricEss(ESS_ID).setPower(new DummyPower(1.0, 0.0, 0.0));
		ess.withSymmetricApplyPowerCallback(rec -> {
			System.out.println(rec.activePower());
		});
		int rechargePower = 50_000;
		new ControllerTest(new ControllerEssPeakShavingImpl()) //
				.addReference("componentManager", new DummyComponentManager()) //
				.addComponent(ess) //
				.addComponent(new DummyElectricityMeter(METER_ID)) //
				.activate(MyConfig.create() //
						.setId(CTRL_ID) //
						.setEssId(ESS_ID) //
						.setMeterId(METER_ID) //
						.setPeakShavingPower(100_000) //
						.setRechargePower(rechargePower) //
						.setMinimumSoc(50) //
						.build())
				.next(new TestCase("Ess in multi use mode") //
						.input(ESS_ACTIVE_POWER, 0) //
						.input(ESS_SOC, 52)
						.input(METER_ACTIVE_POWER, 1_000) //
						.onAfterControllersCallbacks(() -> {
                                                        // simulate follow up controller call
                                                        ess.setActivePower(5);
							assertEquals(0, (int) ess.getActivePower().get());
						})) //
```